### PR TITLE
feat: add get-task-config command

### LINKED ISSUE #182: UCP: retrieve the configuration of the running task from the DM cluster
## Feature Request

**Is your feature request related to a problem? Please describe:**

when the original task config file lost or when debugging the migration task, we may want to retrieve the configuration of the running task from the DM cluster.

**Describe the feature you'd like:**

retrieve the configuration of the running task from the DM cluster

## Score

600

## Mentor(s)
* @csuzhangxc 

Contact the mentors: **#tidb-challenge-program** channel in [TiDB Community](https://join.slack.com/t/tidbcommunity/shared_invite/enQtNzc0MzI4ODExMDc4LWYwYmIzMjZkYzJiNDUxMmZlN2FiMGJkZjAyMzQ5NGU0NGY0NzI3NTYwMjAyNGQ1N2I2ZjAxNzc1OGUwYWM0NzE)  Slack Workspace

## Recommended Skills

* Golang
* Experience with DM

## Learning Materials

- [DM document](https://pingcap.com/docs-cn/stable/reference/tools/data-migration/overview/#dm-%e6%9e%b6%e6%9e%84)
- [DM source reading](https://pingcap.com/blog-cn/#DM-%E6%BA%90%E7%A0%81%E9%98%85%E8%AF%BB)

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve? 
close #182 

### What is changed and how it works?
- save the task-config-string when start-task
- remove the task-config-string when stop-task
- get the task-config-string and write the cfg to file(option)
- add a step by step contribute example

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: dm/master/scheduler/scheduler.go [hunk 1]
```diff
@@ -318,6 +331,124 @@ func (s *Scheduler) RemoveSourceCfg(source string) error {
...
+}
+
+// RemoveTaskCfg removes the config of task
+func (s *Scheduler) RemoveTaskCfg(task string) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// do not remove task if has subtask
+	if _, ok := s.subTaskCfgs[task]; ok {
+		return nil
```
Comment (lance6716) [inline dm/master/scheduler/scheduler.go:424]: or return some error?

File: dm/ctl/master/get_task_config.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	"github.com/spf13/cobra"
+
+	"github.com/pingcap/dm/dm/ctl/common"
+	"github.com/pingcap/dm/dm/pb"
+)
+
+// NewGetTaskCfgCmd creates a getTaskCfg command
+func NewGetTaskCfgCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "get-task-config [task-name] [--file filename]",
```
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply dm/ctl/master/get_task_config.go]: address in e5a307d

File: dm/master/scheduler/scheduler.go (hunk 1 above)
Comment (GMHDBJD) [reply dm/master/scheduler/scheduler.go:424]: We have command `stop-task -s source` for sub-task. So actually it is not a error.

File: step_by_step_contribute.md [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+    make generate_proto
+    ```
+
+3. Generate mock code
+
+    ```sh
+    make generate_mock
+    ```
+
+4. Add some error instance for your new command in [error_list](pkg/terror/error_list.go)
```
Comment (WangXiangUSTC) [inline step_by_step_contribute.md:120]: only when we almost finish the development, can we know what error should be added

File: step_by_step_contribute.md [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+
+    ```sh
+    make unit_test
+    ```
+
+### Step 8: Add Integration Test
+
+1. Add integration test for [dmctl command](tests/dmctl_basic/check_list/get_task_config.sh)
+
+    ```golang
```
Comment (WangXiangUSTC) [inline step_by_step_contribute.md:207]: it is shell

File: dm/ctl/master/get_task_config.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+		return
+	}
+
+	if resp.Result && len(filename) != 0 {
+		file, err := os.Create(filename)
+		defer file.Close()
+		if err != nil {
+			common.PrintLines("can not create file %s:\n%v", filename, err)
+			return
+		}
```
Comment (csuzhangxc) [inline dm/ctl/master/get_task_config.go:70]: How about putting `defer file.Close()` after `if err != nil`? or you can use `ioutil.WriteFile`?

File: dm/master/scheduler/scheduler.go [hunk 6]
```diff
@@ -318,6 +331,126 @@ func (s *Scheduler) RemoveSourceCfg(source string) error {
...
+
+	// do not remove task if has subtask
+	// return nil because we may use `stop-task -s source` for subtask
+	if _, ok := s.subTaskCfgs[task]; ok {
+		return nil
+	}
+
+	if !s.started {
+		return terror.ErrSchedulerNotStarted.Generate()
+	}
```
Comment (csuzhangxc) [inline dm/master/scheduler/scheduler.go:431]: Swap the order of these two blocks?

File: pkg/terror/error_list.go [hunk 7]
```diff
@@ -544,6 +544,8 @@ const (
 	codeSchedulerSubTaskStageInvalidUpdate
 	codeSchedulerSubTaskOpTaskNotExist
 	codeSchedulerSubTaskOpSourceNotExist
+	codeSchedulerTaskExist
```
Comment (csuzhangxc) [inline pkg/terror/error_list.go:547]: not used?

File: step_by_step_contribute.md [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+### Step 2: Clone the forked repository to local storage
+
+    ```sh
+    cd $working_dir # Comes to the directory that you want put the fork in, for example, "cd ~/code"
+    git clone git@github.com:$user/dm.git # Replace "$user" with your GitHub ID
+
+    cd $working_dir/dm
+    git remote add upstream git@github.com:pingcap/dm.git # Adds the upstream repo
+    git remote -v # Confirms that your remote makes sense
+    ```
```
Comment (csuzhangxc) [inline step_by_step_contribute.md:25]: the rendered HTML is strange. and more blocks below.

**Comment by GMHDBJD** [comment]
All comment have been addressed. PTAL again

File: step_by_step_contribute.md [hunk 9]
```diff
@@ -49,21 +49,21 @@ Please perform the following steps to create your Pull Request to this repositor
...
+> This section takes issue [UCP: retrieve the configuration of the running task from the DM cluster](https://github.com/pingcap/dm/issues/182) as an example. Steps of pick up other issues are similar.
 
-Add following comment for the github issue to pick up the challenge program.
+Add following comment for the github issue to pick up the issue.
 
-    ```
-    /pick-up-challenge
-    ```
+```
+/pick-up-issue
```
Comment (csuzhangxc) [inline step_by_step_contribute.md:65]: Can it work?

File: step_by_step_contribute.md [hunk 10]
```diff
@@ -49,21 +49,21 @@ Please perform the following steps to create your Pull Request to this repositor
 
 2. Setting up test environment. See [Test Preparations](tests/README.md#Preparations) for more details.
 
-### Step 4: Pick up a challenge program
+### Step 4: Pick up a issue
 
 You can start by finding an existing issue with the
-[UCP](https://github.com/pingcap/dm/labels/challenge-program-2)
+[help wanted](https://github.com/pingcap/dm/labels/help%20wanted)
```
Comment (csuzhangxc) [inline step_by_step_contribute.md:55]: If you want, you can add this label to UCP issues.

File: step_by_step_contribute.md (hunk 9 above)
Comment (GMHDBJD) [reply step_by_step_contribute.md]: remove it in 1a13556

File: step_by_step_contribute.md (hunk 10 above)
Comment (GMHDBJD) [reply step_by_step_contribute.md:55]: Done.

**Pull request merged**